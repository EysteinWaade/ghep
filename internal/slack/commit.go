package slack

import (
	"bytes"
	"fmt"

	"github.com/navikt/ghep/internal/github"
)

func CreateCommitMessage(channel string, event github.CommitEvent) ([]byte, error) {
	type commit struct {
		URL     string
		Ref     string
		Message string
	}

	type text struct {
		Channel    string
		URL        string
		Repository string
		Author     string
		Commits    []commit
	}

	payload := text{
		Channel:    channel,
		URL:        event.Repository.HtmlUrl,
		Author:     event.Commits[0].Author.Name,
		Repository: event.Repository.Name,
	}

	commits := []commit{}
	for _, c := range event.Commits {
		commits = append(commits, commit{
			Ref:     c.Id[:8],
			Message: c.Message,
			URL:     c.Url,
		})
	}
	payload.Commits = commits

	var output bytes.Buffer
	if err := commitTmpl.Execute(&output, payload); err != nil {
		return nil, fmt.Errorf("executing commit template: %w", err)
	}

	return output.Bytes(), nil
}
