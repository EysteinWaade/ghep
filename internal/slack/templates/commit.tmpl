{
  "channel": "{{ .Channel }}",
  "text": "<{{ .Compare }}|{{ .Commits | len }} new commits> pushed to `<{{ .URL }}|{{ .Repository }}>` by {{ .Author }}",
  "attachments": [
     {
       "text": "{{ range .Commits }}`<{{ .URL }}|{{ .Ref }}>` - {{ .Message }}\n{{ end }}",
       "color": "#000"
     }
   ]
}
