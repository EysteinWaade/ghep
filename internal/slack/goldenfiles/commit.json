{
  "channel": "#test",
  "text": "1 new commits pushed to `<https://github.com/test|test>` by Ola Nordmann",
  "attachments": [
     {
       "text": "`<https://github.com/test|d6f21c84>` - test\n"
     }
   ]
}
